# Лабораторная работа №1
## Анализ пассажиров корабля "Титаник"

**Заполните недостающий код ("You code here")**

In [4]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
# Graphics in SVG format are more sharp and legible
%config InlineBackend.figure_format = 'svg'
pd.set_option("display.precision", 2)

** Чтение данных в Pandas DataFrame**

In [5]:
data = pd.read_csv('titanic_train.csv',
                  index_col='PassengerId')

**Первые 5 строк**

In [6]:
data.head(5)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket   Fare Cabin Embarked  
PassengerId                                                        
1                1      0         A/5 21171   7.25   NaN        S  
2                1      0          PC 17599  71.28   C85        C  
3                0      0  STON/O2. 3101282   7.92   NaN        S  
4                1      0            113803  53.10  C123        S  
5                0      0            373450   8.05   NaN        S

In [7]:
data.describe()

Survived  Pclass     Age   SibSp   Parch    Fare
count    891.00  891.00  714.00  891.00  891.00  891.00
mean       0.38    2.31   29.70    0.52    0.38   32.20
std        0.49    0.84   14.53    1.10    0.81   49.69
min        0.00    1.00    0.42    0.00    0.00    0.00
25%        0.00    2.00   20.12    0.00    0.00    7.91
50%        0.00    3.00   28.00    0.00    0.00   14.45
75%        1.00    3.00   38.00    1.00    0.00   31.00
max        1.00    3.00   80.00    8.00    6.00  512.33

**Давайте выберем тех пассажиров, которые отправились в Cherbourg (Embarked = C) и заплатили > 200 фунтов за билет (fare > 200).**

Убедитесь, что вы понимаете, как на самом деле работает эта конструкция.

In [8]:
data[(data['Embarked'] == 'C') & (data.Fare > 200)].head()

Survived  Pclass  \
PassengerId                     
119                 0       1   
259                 1       1   
300                 1       1   
312                 1       1   
378                 0       1   

                                                        Name     Sex   Age  \
PassengerId                                                                  
119                                 Baxter, Mr. Quigg Edmond    male  24.0   
259                                         Ward, Miss. Anna  female  35.0   
300          Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0   
312                               Ryerson, Miss. Emily Borie  female  18.0   
378                                Widener, Mr. Harry Elkins    male  27.0   

             SibSp  Parch    Ticket    Fare            Cabin Embarked  
PassengerId                                                            
119              0      1  PC 17558  247.52          B58 B60        C  
259              0      0  PC 17755  512.33              NaN        C  
300              0      1  PC 17558  247.52          B58 B60        C  
312              2      2  PC 17608  262.38  B57 B59 B63 B66        C  
378              0      2    113503  211.50              C82        C

** Мы можем отсортировать этих людей по тарифу (Fare) в порядке убывания. **

In [9]:
data[(data['Embarked'] == 'C') & 
     (data['Fare'] > 200)].sort_values(by='Fare',
                               ascending=False).head()

Survived  Pclass                                   Name     Sex  \
PassengerId                                                                    
259                 1       1                       Ward, Miss. Anna  female   
680                 1       1     Cardeza, Mr. Thomas Drake Martinez    male   
738                 1       1                 Lesurer, Mr. Gustave J    male   
312                 1       1             Ryerson, Miss. Emily Borie  female   
743                 1       1  Ryerson, Miss. Susan Parker "Suzette"  female   

              Age  SibSp  Parch    Ticket    Fare            Cabin Embarked  
PassengerId                                                                  
259          35.0      0      0  PC 17755  512.33              NaN        C  
680          36.0      0      1  PC 17755  512.33      B51 B53 B55        C  
738          35.0      0      0  PC 17755  512.33             B101        C  
312          18.0      2      2  PC 17608  262.38  B57 B59 B63 B66        C  
743          21.0      2      2  PC 17608  262.38  B57 B59 B63 B66        C

**Давайте добавим новый признак.**

In [10]:
def age_category(age):
    '''
    < 30 -> 1
    >= 30, <55 -> 2
    >= 55 -> 3
    '''
    if age < 30:
        return 1
    elif age < 55:
        return 2
    elif age >= 55:
        return 3

In [11]:
age_categories = [age_category(age) for age in data.Age]
data['Age_category'] = age_categories

**Другой способ сделать это с `apply`.**

In [12]:
data['Age_category'] = data['Age'].apply(age_category)

In [13]:
print(data['Sex'].value_counts())


Sex
male      577
female    314
Name: count, dtype: int64


**1. Сколько мужчин / женщин было на борту?**
- 412 мужчин и 479 женщин
- 314 мужчин и 577 женщин
- 479 мужчин и 412 женщин
- 577 мужчин и 314 женщин

In [14]:

men = data[(data['Sex'] == 'male') & (data['Pclass'] == 2)]
print("Мужчин во 2 классе:", len(men)) 
women = data[(data['Sex'] == 'female') & (data['Pclass'] == 2)]
print("Женщин во 2 классе:", len(women))


Мужчин во 2 классе: 108
Женщин во 2 классе: 76


**2. Определите распределение функции `Pclass`. Теперь Для мужчин и женщин отдельно. Сколько людей из второго класса было на борту?**
- 104
- 108
- 112
- 125

In [15]:
print("Медиана:", round(data['Fare'].median(), 2))
print("Стандартное отклонение:", round(data['Fare'].std(), 2))

Медиана: 14.45
Стандартное отклонение: 49.69


**3. Каковы медиана и стандартное отклонение`Fare`?. Округлите до 2-х знаков после запятой.**
- медиана 14,45, стандартное отклонение 49,69
- медиана 15,1, стандартное отклонение 12,15
- медиана 13,15, стандартное отклонение 35,3
- Медиана 17,43, стандартное отклонение - 39,1.

In [16]:
survived_age = data[data['Survived'] == 1]['Age'].mean()
died_age = data[data['Survived'] == 0]['Age'].mean()
print("Средний возраст выживших:", round(survived_age, 2))
print("Средний возраст", round(died_age, 2))
print(survived_age > died_age)

Средний возраст выживших: 28.34
Средний возраст 30.63
False


** 4. Правда ли, что средний возраст выживших людей выше, чем у пассажиров, которые в конечном итоге умерли? **
- Да
- Нет


In [17]:
young = data[data['Age'] < 30]
old = data[data['Age'] > 60]
young_survived = young[young['Survived'] == 1]
old_survived = old[old['Survived'] == 1]

print("Молодые:", round(len(young_survived)/len(young)*100, 1), "%")
print("Пожилые:", round(len(old_survived)/len(old)*100, 1), "%")

Молодые: 40.6 %
Пожилые: 22.7 %


**5. Это правда, что пассажиры моложе 30 лет. выжили чаще, чем те, кому больше 60 лет. Каковы доли выживших людей среди молодых и пожилых людей?**
- 22,7% среди молодежи и 40,6% среди пожилых
- 40,6% среди молодежи и 22,7% среди пожилых
- 35,3% среди молодежи и 27,4% среди пожилых
- 27,4% среди молодежи и 35,3% среди пожилых

In [18]:
men = data[data['Sex'] == 'male']
women = data[data['Sex'] == 'female']
men_survived = men[men['Survived'] == 1]
women_survived = women[women['Survived'] == 1]

print("Мужчины:", round(len(men_survived)/len(men)*100, 1), "%")
print("Женщины:", round(len(women_survived)/len(women)*100, 1), "%")

Мужчины: 18.9 %
Женщины: 74.2 %


**6. Правда ли, что женщины выживали чаще мужчин? Каковы доли выживших людей среди мужчин и женщин?**
- 30,2% среди мужчин и 46,2% среди женщин
- 35,7% среди мужчин и 74,2% среди женщин
- 21,1% среди мужчин и 46,2% среди женщин
- 18,9% среди мужчин и 74,2% среди женщин

In [19]:
john_count = 0
william_count = 0
thomas_count = 0
charles_count = 0
male_names = data[data['Sex'] == 'male']['Name']
for name in male_names:
    if 'John' in name:
        john_count += 1
    if 'William' in name:
        william_count += 1
    if 'Thomas' in name:
        thomas_count += 1
    if 'Charles' in name:
        charles_count += 1

print("John:", john_count)
print("William:", william_count)
print("Thomas:", thomas_count)
print("Charles:", charles_count)

John: 41
William: 54
Thomas: 16
Charles: 20


**7. Какое имя наиболее популярно среди пассажиров мужского пола?**
- Charles
- Thomas
- William
- John

In [20]:
age_by_class = data.groupby(['Pclass', 'Sex'])['Age'].mean()
print(round(age_by_class, 1))

Pclass  Sex   
1       female    34.6
        male      41.3
2       female    28.7
        male      30.7
3       female    21.8
        male      26.5
Name: Age, dtype: float64


**8. Как средний возраст мужчин / женщин зависит от `Pclass`? Выберите все правильные утверждения:**
- В среднем мужчины 1 класса старше 40 лет
- В среднем женщины 1 класса старше 40 лет
- Мужчины всех классов в среднем старше, чем женщины того же класса
- В среднем, пассажиры первого класса старше, чем пассажиры 2-го класса, которые старше, чем пассажиры 3-го класса.